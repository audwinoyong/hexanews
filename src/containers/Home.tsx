import * as React from 'react'
import Grid from '@material-ui/core/Grid'
import Divider from '@material-ui/core/Divider'
import Typography from '@material-ui/core/Typography'
import { History } from 'history'

import ArticleCard from '../components/ArticleCard'
import ArticleRow from '../components/ArticleRow'
import PopularArticleRow from '../components/PopularArticleRow'

type Props = {
  history: History
}

const styles = {
  articleDivider: {
    marginTop: 20,
    marginLeft: 24,
  },
  container: {
    padding: '5%',
    paddingLeft: '7%',
  },
  divider: {
    marginTop: 64,
    marginRight: 18,
  },
  latestPostContainer: {
    marginRight: 24,
  },
  popularArticles: {
    marginLeft: 36,
  },
  topStory: {
    marginTop: 32,
    marginLeft: 24,
  },
}

export default class Home extends React.Component<Props> {

  public render() {
    return (
      <div style={styles.container}>
        <Grid container spacing={24}>
          <Grid item xs={4}>
            <ArticleCard
              history={this.props.history}
              title='Adjustable sidebar using Angular'
              description='Also for this tutorial I din’t put accent on CSS,
                I assume that you already created a sidebar which you need to make draggable.'
            />
          </Grid>
          <Grid item xs={4}>
            <ArticleCard
              history={this.props.history}
              title='Implement Google Analytics for React Native'
              description='For almost every product that is built or launched in the market, the business and development
               teams testing there'
            />
          </Grid>
          <Grid item xs={4}>
            <ArticleCard
              history={this.props.history}
              title='One Book at The Time —A Short Story'
              description='There had been times she had almost given up. All the lying, the sneaking around.
                But today she once more managed to continue'
            />
          </Grid>
        </Grid>
        <Divider style={styles.divider}/>
        <Grid container spacing={24}>
          <Grid item xs={8} style={styles.latestPostContainer}>
            <ArticleRow
              title='We Still Don’t Know Whether Uber Is a Real Business'
              description='It has never had to live on the cash it generates.'
            />
            <ArticleRow
              title='Happy All the Time'
              description='As biometric tracking takes over the modern workplace'
            />
            <ArticleRow
              title='If you’re a developer, you should start blogging — and here’s why'
              description='My blogging journey and skills I’ve acquired along the way'
            />
          </Grid>
          <Grid item xs={3}>
            <Typography variant='headline' component='h2' style={styles.topStory}>
              Top story in Hexanews
            </Typography>
            <Divider style={styles.articleDivider}/>
            <div style={styles.popularArticles}>
              <PopularArticleRow
                author={'Daniel Wu'}
                title={'The software engineer’s guide'}
              />
              <PopularArticleRow
                author={'Jesse Weaver'}
                title={'Design won\'t save the world'}
              />
              <PopularArticleRow
                author={'Mohammed Aladdin'}
                title={'Write clean code and get rid of code smell'}
              />
            </div>
          </Grid>
        </Grid>
      </div>
    )
  }
}
