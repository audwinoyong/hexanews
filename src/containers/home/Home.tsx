/**
 * Home container for displaying all articles and featured articles
 */

import * as React from 'react'
import { connect } from 'react-redux'
import { Dispatch } from 'redux'
import Grid from '@material-ui/core/Grid'
import Divider from '@material-ui/core/Divider'
import Typography from '@material-ui/core/Typography'
import { map, orderBy, take, find, filter } from 'lodash'

import ArticleCard from '../../components/articleCard/ArticleCard'
import ArticleRow from '../../components/articleRow/ArticleRow'
import PopularArticleRow from '../../components/popularArticleRow/PopularArticleRow'
import styles from './styles'
import CategoryHeader from '../../components/categoryHeader/CategoryHeader'
import { homeString } from '../../constants/string'
import selector, { StateProps } from './selector'
import { User } from '../../domain/model/User'

type Props = {
  dispatch: Dispatch<any>
} & StateProps

export class Home extends React.Component<Props> {

  renderAllArticles = () => {
    const { articles, dispatch, users } = this.props
    const orderedArticles = orderBy(articles, ['createdAt'], ['desc'])
    return map(orderedArticles, (article) => {
      const author = find(users, (user: User) => user._id === article.authorId)
      const authorName = author ? author.name : ''
      return (
        <ArticleRow
          article={article}
          authorName={authorName}
          dispatch={dispatch}
          key={article._id}
        />
      )
    })
  }

  renderFeaturedArticles = () => {
    const { articles, dispatch } = this.props
    const featuredArticles = filter(articles, ['isFeatured', true])
    return map(featuredArticles, (article) => {
      return (
        <Grid item md={4} xs={12} key={article._id}>
          <ArticleCard
            article={article}
            dispatch={dispatch}
            key={article._id}
          />
        </Grid>
      )
    })
  }

  renderPopularArticles = () => {
    const { articles, dispatch, users } = this.props
    const sortedPopularArticles = orderBy(articles, ['views'], ['desc'])
    const threePopularArticles = take(sortedPopularArticles, 3)
    return map(threePopularArticles, (article) => {
      const author = find(users, (user: User) => user._id === article.authorId)
      const authorName = author ? author.name : ''
      return (
        <PopularArticleRow
          article={article}
          authorName={authorName}
          dispatch={dispatch}
          key={article._id}
        />
      )
    })
  }

  public render() {
    return (
      <div style={styles.container}>
        <CategoryHeader />
        <Grid container spacing={24} style={styles.gridContainer}>
          {this.renderFeaturedArticles()}
        </Grid>
        <Divider style={styles.divider} />
        <Grid container spacing={24}>
          <Grid item md={8} xs={12} style={styles.latestArticleContainer}>
            {this.renderAllArticles()}
          </Grid>
          <Grid item md={3} xs={12}>
            <Typography variant='headline' component='h2' style={styles.topStory}>
              {homeString.topStory}
            </Typography>
            <Divider style={styles.articleDivider} />
            <div style={styles.popularArticles}>
              {this.renderPopularArticles()}
            </div>
          </Grid>
        </Grid>
      </div>
    )
  }
}

export default connect(selector)(Home)