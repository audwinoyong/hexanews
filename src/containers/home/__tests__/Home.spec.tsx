/**
 * Unit tests for checking the behavior of Home container
 */

import * as React from 'react'
import * as renderer from 'react-test-renderer'
import * as Enzyme from 'enzyme'
import { BrowserRouter } from 'react-router-dom'
import * as Adapter from 'enzyme-adapter-react-16'

import { Home } from '../Home'

Enzyme.configure({ adapter: new Adapter() })

const mockUser = {
  _id: '1',
  createdAt: 'created',
  description: 'desc',
  email: 'user@email.com',
  name: 'user'
}

describe('Home', () => {

  it('renders Home component', () => {
    const component = renderer.create(
      <BrowserRouter>
        <Home
          articles={[]}
          categories={[]}
          dispatch={jest.fn()}
          users={[mockUser]}
        />
      </BrowserRouter>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})