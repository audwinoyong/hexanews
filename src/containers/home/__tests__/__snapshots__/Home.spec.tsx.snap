// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders Home component 1`] = `
<div
  className=""
  style={
    Object {
      "padding": "5%",
      "paddingLeft": "6%",
      "paddingTop": "0%",
    }
  }
>
  <div
    className=""
    style={
      Object {
        "flexGrow": 1,
        "overflow": "auto",
      }
    }
  >
    <header
      className="MuiPaper-root-106 MuiPaper-elevation4-112 MuiAppBar-root-98 MuiAppBar-positionStatic-102 MuiAppBar-colorPrimary-104"
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "boxShadow": "none",
          "paddingRight": 18,
        }
      }
    >
      <div
        className="MuiToolbar-root-133 MuiToolbar-regular-135 MuiToolbar-gutters-134"
      />
    </header>
  </div>
  <div
    className="MuiGrid-container-1 MuiGrid-spacing-xs-24-25"
    style={
      Object {
        "paddingTop": "2%",
      }
    }
  />
  <hr
    className="MuiDivider-root-137"
    style={
      Object {
        "marginRight": 18,
        "marginTop": 64,
      }
    }
  />
  <div
    className="MuiGrid-container-1 MuiGrid-spacing-xs-24-25"
  >
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 MuiGrid-grid-md-8-65"
      style={
        Object {
          "marginRight": 24,
        }
      }
    />
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 MuiGrid-grid-md-3-60"
    >
      <h2
        className="MuiTypography-root-141 MuiTypography-title-147"
        style={
          Object {
            "fontWeight": "bold",
            "marginLeft": 24,
            "marginTop": 32,
          }
        }
      >
        Top story of Hexanews
      </h2>
      <hr
        className="MuiDivider-root-137"
        style={
          Object {
            "marginLeft": 24,
            "marginTop": 20,
          }
        }
      />
      <div
        className=""
        style={
          Object {
            "marginLeft": 36,
          }
        }
      />
    </div>
  </div>
</div>
`;
